### PR TITLE
hash: use generic instantiation

### PR DESCRIPTION
**Pull request by howardjohn**
This enables doing something like
https://github.com/opencontainers/go-digest/issues/71#issuecomment-1330069987 to override the hashing method (example: replace with sha256simd).

When an importer does not overriden, the library should behave exactly the same.

See https://github.com/google/go-containerregistry/issues/1330

### CODE REVIEW

**Comment by jonjohnsonjr** [comment]
Oh that's wonderful. A few more places:

```
registry/manifest.go
113:		rd := sha256.Sum256(m.blob)
140:		rd := sha256.Sum256(m.blob)
156:		rd := sha256.Sum256(b.Bytes())

registry/registry_test.go
50:	h := sha256.Sum256([]byte(s))
```

```
legacy/tarball/write.go
66:	rawDigest := sha256.Sum256([]byte(s))
```

**Comment by howardjohn** [comment]
> legacy/tarball/write.go

Thanks, missed those ones! Updated a bit. Note the ignored errors match what the old function did internally so I think its safe

**Review by jonjohnsonjr — APPROVED** [review]

File: pkg/registry/registry_test.go [hunk 1]
```diff
@@ -47,8 +47,8 @@ const (
 )
 
 func sha256String(s string) string {
-	h := sha256.Sum256([]byte(s))
-	return hex.EncodeToString(h[:])
+	h, _, _ := v1.SHA256(bytes.NewReader([]byte(s)))
```
Comment (imjasonh) [inline pkg/registry/registry_test.go:50]: strings.NewReader could be better here

**Comment by imjasonh** [comment]
We might want a lint check that `crypto/sha256` doesn't sneak back in in a future change, and to steer folks toward `crypto.SHA256`. I know I'm going to forget about this, old habits die hard.

**Pull request merged**